File: ios/Anemone SDK/general/client/HttpRequest.swift
//
//  HttpRequest.swift
//  Anemone SDK
//
//  Created by Angel Luis Garcia on 25/08/2017.
//  Copyright © 2017 Syltek Solutions S.L. All rights reserved.
//

import Foundation

public struct HttpRequest {
    public let method: HttpMethod
    public let url: String
    public let queryParams: [String: Any]?
    public let bodyParams: Any?
    public let headers: [String: String]?
    public let timeout: TimeInterval?

    private init(method: HttpMethod, url: String, queryParams: [String: Any]?, bodyParams: Any?, headers: [String: String]?, timeout: TimeInterval?) {
        self.method = method
        self.url = url
        self.queryParams = queryParams
        self.bodyParams = bodyParams
        self.headers = headers
        self.timeout = timeout
    }

    public func copy(
        method: HttpMethod? = nil,
        url: String? = nil,
        queryParams: [String: Any]? = nil,
        bodyParams: [String: Any]? = nil,
        headers: [String: String]? = nil,
        timeout: TimeInterval? = nil
    ) -> HttpRequest {
        HttpRequest(
            method: method ?? self.method,
            url: url ?? self.url,
            queryParams: queryParams ?? self.queryParams,
            bodyParams: bodyParams ?? self.bodyParams,
            headers: headers ?? self.headers,
            timeout: timeout ?? self.timeout
        )
    }

    public var contentType: String? {
        headers?["Content-Type"]
    }
}

public extension HttpRequest {
    static let includeTotalHeaderKey: String = "X-Playtomic-Include-Total-Header"
    static let includeTotalHeaderValue: String = "true"

    static func get(url: String, params: [String: Any]?, timeout: TimeInterval? = nil) -> HttpRequest {
        HttpRequest(method: .get, url: url, queryParams: params, bodyParams: nil, headers: nil, timeout: timeout)
    }

    static func post(url: String, params: Any?, timeout: TimeInterval? = nil, headers: [String: String]? = nil) -> HttpRequest {
        HttpRequest(method: .post, url: url, queryParams: nil, bodyParams: params, headers: headers, timeout: timeout)
    }

    static func post(
        url: String,
        multipartBodyParams: [MultipartBodyParam],
        multipartDataFiles: [MultipartDataFile],
        timeout: TimeInterval? = nil
    ) -> HttpRequest {
        let multipartBody = MultipartBody(
            multipartBodyParams: multipartBodyParams,
            files: multipartDataFiles
        )
        return HttpRequest(
            method: .post,
            url: url,
            queryParams: nil,
            bodyParams: multipartBody,
            headers: ["Content-Type": "multipart/form-data; boundary=\(multipartBody.boundary)"],
            timeout: timeout
        )
    }

    static func put(url: String, params: Any?, timeout: TimeInterval? = nil) -> HttpRequest {
        HttpRequest(method: .put, url: url, queryParams: nil, bodyParams: params, headers: nil, timeout: timeout)
    }
    
    static func put(
        url: String,
        multipartBodyParams: [MultipartBodyParam],
        multipartDataFiles: [MultipartDataFile],
        timeout: TimeInterval? = nil
    ) -> HttpRequest {
        let multipartBody = MultipartBody(
            multipartBodyParams: multipartBodyParams,
            files: multipartDataFiles
        )
        return HttpRequest(
            method: .put,
            url: url,
            queryParams: nil,
            bodyParams: multipartBody,
            headers: ["Content-Type": "multipart/form-data; boundary=\(multipartBody.boundary)"],
            timeout: timeout
        )
    }
    
    static func patch(
        url: String,
        multipartBodyParams: [MultipartBodyParam],
        multipartDataFiles: [MultipartDataFile],
        timeout: TimeInterval? = nil
    ) -> HttpRequest {
        let multipartBody = MultipartBody(
            multipartBodyParams: multipartBodyParams,
            files: multipartDataFiles
        )
        return HttpRequest(
            method: .patch,
            url: url,
            queryParams: nil,
            bodyParams: multipartBody,
            headers: ["Content-Type": "multipart/form-data; boundary=\(multipartBody.boundary)"],
            timeout: timeout
        )
    }

    static func patch(url: String, params: Any?, timeout: TimeInterval? = nil) -> HttpRequest {
        HttpRequest(method: .patch, url: url, queryParams: nil, bodyParams: params, headers: nil, timeout: timeout)
    }

    static func delete(url: String, params: [String: Any]?, timeout: TimeInterval? = nil) -> HttpRequest {
        HttpRequest(method: .delete, url: url, queryParams: params, bodyParams: nil, headers: nil, timeout: timeout)
    }

    static func head(url: String, params: [String: Any]?, headers: [String: String]? = nil, timeout: TimeInterval? = nil) -> HttpRequest {
        HttpRequest(method: HttpMethod.head, url: url, queryParams: params, bodyParams: nil, headers: headers, timeout: timeout)
    }
}

extension HttpRequest: Equatable { }
public func == (lhs: HttpRequest, rhs: HttpRequest) -> Bool {
    let lhsBody = (lhs.bodyParams as? [String: Any])?.description
        ?? (lhs.bodyParams as? [[String: Any]])?.description
        ?? (lhs.bodyParams as? MultipartBody)?.boundary
    let rhsBody = (rhs.bodyParams as? [String: Any])?.description
        ?? (rhs.bodyParams as? [[String: Any]])?.description
        ?? (rhs.bodyParams as? MultipartBody)?.boundary
    return lhs.method == rhs.method &&
        lhs.url == rhs.url &&
        lhs.queryParams?.description == rhs.queryParams?.description &&
        lhsBody == rhsBody &&
        (lhs.headers ?? [:]) == (rhs.headers ?? [:]) &&
        lhs.timeout == rhs.timeout
}

File: android/app/src/main/java/com/playtomic/test/anemone_sdk/general/client/HttpRequest.kt
package com.anemonesdk.general.client

/**
 * Created by manuelgonzalezvillegas on 25/8/17.
 */

data class HttpRequest private constructor(
    val method: HttpMethod,
    val url: String,
    val queryParams: Map<String, Any>?,
    val bodyParams: Any?,
    val headers: Map<String, String>?,
    val timeout: Long?
) {

    val contentType get() = headers?.get("Content-Type")

    companion object {
        val includeTotalHeaderKey: String = "X-Playtomic-Include-Total-Header"
        val includeTotalHeaderValue: String = "true"

        fun get(url: String, params: Map<String, Any>?, timeout: Long? = null): HttpRequest =
            HttpRequest(method = HttpMethod.get, url = url, queryParams = params, bodyParams = null, headers = null, timeout = timeout)

        fun post(url: String, params: Any?, timeout: Long? = null): HttpRequest =
            HttpRequest(method = HttpMethod.post, url = url, queryParams = null, bodyParams = params, headers = null, timeout = timeout)

        fun post(url: String, params: Any?, headers: Map<String, String>?): HttpRequest =
            HttpRequest(method = HttpMethod.post, url = url, queryParams = null, bodyParams = params, headers = headers, timeout = null)

        fun post(url: String, multipartBodyParams: List<MultipartBodyParam>, multipartDataFiles: List<MultipartDataFile>, timeout: Long? = null): HttpRequest =
            HttpRequest(
                method = HttpMethod.post,
                url = url,
                queryParams = null,
                bodyParams = MultipartBody(multipartBodyParams, multipartDataFiles),
                headers = mapOf(Pair("Content-Type", "multipart/form-data")),
                timeout = timeout
            )

        fun put(url: String, params: Any?, timeout: Long? = null): HttpRequest =
            HttpRequest(method = HttpMethod.put, url = url, queryParams = null, bodyParams = params, headers = null, timeout = timeout)

        fun put(url: String, multipartBodyParams: List<MultipartBodyParam>, multipartDataFiles: List<MultipartDataFile>, timeout: Long? = null): HttpRequest =
            HttpRequest(
                method = HttpMethod.put,
                url = url,
                queryParams = null,
                bodyParams = MultipartBody(multipartBodyParams, multipartDataFiles),
                headers = mapOf(Pair("Content-Type", "multipart/form-data")),
                timeout = timeout
            )

        fun patch(url: String, params: Any?, timeout: Long? = null): HttpRequest =
            HttpRequest(method = HttpMethod.patch, url = url, queryParams = null, bodyParams = params, headers = null, timeout = timeout)

        fun patch(url: String, multipartBodyParams: List<MultipartBodyParam>, multipartDataFiles: List<MultipartDataFile>, timeout: Long? = null): HttpRequest =
            HttpRequest(
                method = HttpMethod.patch,
                url = url,
                queryParams = null,
                bodyParams = MultipartBody(multipartBodyParams, multipartDataFiles),
                headers = mapOf(Pair("Content-Type", "multipart/form-data")),
                timeout = timeout
            )

        fun delete(url: String, params: Map<String, Any>?, timeout: Long? = null): HttpRequest =
            HttpRequest(method = HttpMethod.delete, url = url, queryParams = params, bodyParams = null, headers = null, timeout = timeout)

        fun head(url: String, params: Map<String, Any>?, headers: Map<String, String>? = null, timeout: Long? = null): HttpRequest =
            HttpRequest(method = HttpMethod.head, url = url, queryParams = params, bodyParams = null, headers = headers , timeout = timeout)
    }
}
